src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -22,47 +31,131 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics
     /// Each of these taggers is nicely asynchronous and properly works within the async
     /// tagging infrastructure. 
     /// </summary>
-    internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> :
-        ForegroundThreadAffinitizedObject,
-        ITaggerProvider
+    internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> : AsynchronousTaggerProvider<TTag>
         where TTag : ITag
     {
-        private readonly object _uniqueKey = new object();
         private readonly IDiagnosticService _diagnosticService;
-        private readonly IForegroundNotificationService _notificationService;
-        private readonly IAsynchronousOperationListener _listener;
 
         protected AbstractDiagnosticsTaggerProvider(
             IDiagnosticService diagnosticService,
             IForegroundNotificationService notificationService,
             IAsynchronousOperationListener listener)
+            : base(listener, notificationService)
         {
             _diagnosticService = diagnosticService;
-            _notificationService = notificationService;
-            _listener = listener;
         }
 
-        protected internal abstract IEnumerable<Option<bool>> Options { get; }
+        protected override TaggerDelay AddedTagNotificationDelay => TaggerDelay.OnIdle;
+
+        protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            return TaggerEventSources.Compose(
```

Comment:
I would add ActiveContextChanged Event source here as well. and remove this (http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs,58)

looks like we added the code above as a quick way to handle context change when it is first introduced to workspace. but that is not right way to handle it. DiagnosticAnalyzerService shouldn't care about buffer. it should only care about solution.

Reply:
hrmm.. can a buffer even cross workspaces though?  i think it can (from Misc Workspace to normal workspace, and vice versa).  So i think we need WorkspaceRegistrationChanged as well.  I will add ActiveContextChanged though.